### PR TITLE
Remote reindex: Add support for configurable retry mechanism

### PR DESCRIPTION
**Pull request by ankitkala**
### Description
Remote Reindex mostly involves repeating 2 major steps: 
1. fetch data using scroll call. 
2. Ingest the data fetched using a transport level bulk request.

I've added following minor enhancements to the remote reindex process.
- Current retry mechanism retries the entire reindex process from the start. I've changed the behavior of retry logic to scroll requests to remote cluster (Bulk already has a retry mechanism).
- Current retry mechanism would only kick in for 429s. I've also added retried for 5xx & ConnectException.
- Exposed the retry params behind cluster settings. 
- Also, changed the logger to prefix based. This is useful to segregate logs from multiple ongoing reindexes for easier debugging. 
- Added debug logs for few places for ease of tracking the reindex operation.
 
### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/12560

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by ankitkala** [comment]
I'm still working on adding tests for this change. Wanted to get the changes out for initial feedback while i work on tests.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 1]
```diff
@@ -180,17 +193,31 @@ protected void cleanup(Runnable onCompletion) {
...
     }
 
+    private void executeWithRetries(
+        Request request,
+        BiFunction<XContentParser, MediaType, Response> parser,
+        RejectAwareActionListener<Response> childListener
+    ) {
+        execute(request, parser,
+            new RetryListener(logger, threadPool, backoffPolicy, r -> {
+                logger.info("Retrying execute request");
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: Oops. Removed the commented block.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 2]
```diff
@@ -124,12 +135,14 @@ private void onStartResponse(RejectAwareActionListener<Response> searchListener,
 
     @Override
     protected void doStartNextScroll(String scrollId, TimeValue extraKeepAlive, RejectAwareActionListener<Response> searchListener) {
+        logger.debug("Starting next scroll call");
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: Sure. Moved all additional logging to trace.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 3]
```diff
@@ -180,17 +193,31 @@ protected void cleanup(Runnable onCompletion) {
...
         RejectAwareActionListener<? super T> listener
     ) {
+        logger.debug("Executing http request to remote cluster");
         // Preserve the thread context so headers survive after the call
         java.util.function.Supplier<ThreadContext.StoredContext> contextSupplier = threadPool.getThreadContext().newRestorableContext(true);
         try {
             client.performRequestAsync(request, new ResponseListener() {
                 @Override
                 public void onSuccess(org.opensearch.client.Response response) {
+                    logger.debug("Successfully got response from the remote");
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: done.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 4]
```diff
@@ -180,17 +193,31 @@ protected void cleanup(Runnable onCompletion) {
...
+                execute(request, parser, r);
+            },  childListener));
+    }
+
     private <T> void execute(
         Request request,
         BiFunction<XContentParser, MediaType, T> parser,
         RejectAwareActionListener<? super T> listener
     ) {
+        logger.debug("Executing http request to remote cluster");
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: done

File: server/src/main/java/org/opensearch/index/reindex/RetryListener.java [hunk 5]
```diff
@@ -84,7 +84,7 @@ public void onRejection(Exception e) {
         if (retries.hasNext()) {
             retryCount += 1;
             TimeValue delay = retries.next();
-            logger.trace(() -> new ParameterizedMessage("retrying rejected search after [{}]", delay), e);
+            logger.info(() -> new ParameterizedMessage("retrying rejected search after [{}]", delay, e.getMessage()));
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/reindex/RetryListener.java]: moved back to trace

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 6]
```diff
@@ -237,22 +264,27 @@ public void onSuccess(org.opensearch.client.Response response) {
...
                             ResponseException re = (ResponseException) e;
                             int statusCode = re.getResponse().getStatusLine().getStatusCode();
                             e = wrapExceptionToPreserveStatus(statusCode, re.getResponse().getEntity(), re);
-                            if (RestStatus.TOO_MANY_REQUESTS.getStatus() == statusCode) {
+                            // retry all 5xx & 429s.
+                            if (RestStatus.TOO_MANY_REQUESTS.getStatus() == statusCode || statusCode >= RestStatus.INTERNAL_SERVER_ERROR.getStatus()) {
                                 listener.onRejection(e);
                                 return;
                             }
+                        } else if (e instanceof ConnectException) {
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java:277]: It depends completely on the underlying http client. We just need to backoff & retry.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java [hunk 7]
```diff
@@ -99,21 +102,29 @@ public RemoteScrollableHitSource(
...
+        // Skipping searchListener::onRejection(used for retries) for remote source as we've configured retries at request(scroll) level.
+        }, searchListener::onFailure, searchListener::onFailure));
     }
 
     void lookupRemoteVersion(RejectAwareActionListener<Version> listener) {
-        execute(new Request("GET", ""), MAIN_ACTION_PARSER, listener);
+        logger.debug("Checking version for remote domain");
+        // We're skipping retries for the first call to remote cluster so that we fail fast & respond back immediately
+        // instead of retrying for longer duration.
+        //execute(new Request("GET", ""), MAIN_ACTION_PARSER, listener);
```
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: Updated.

File: modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankitkala) [reply modules/reindex/src/main/java/org/opensearch/index/reindex/remote/RemoteScrollableHitSource.java]: ack

**Comment by ankitkala** [comment]
> ❕ Gradle check result for [9e2fc06](https://github.com/opensearch-project/OpenSearch/commit/9e2fc060dc26a0db3984783d5cc094e72030ba96): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/34703/)
> 
> * **TEST FAILURES:**
> 
> ```
>       1 org.opensearch.search.SearchWeightedRoutingIT.testMultiGetWithNetworkDisruption_FailOpenEnabled
> ```
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

Flaky test: https://github.com/opensearch-project/OpenSearch/issues/10755

**Comment by ankitkala** [comment]
> MinimumClusterManagerNodesIT

Flaky test: https://github.com/opensearch-project/OpenSearch/issues/10006

**Review by shwetathareja — APPROVED** [review]

**Comment by ankitkala** [comment]
Issue for documentation update: https://github.com/opensearch-project/documentation-website/issues/6674

**Pull request merged**